pkg/tarsplitutils/tarstream.go

```diff
@@ -0,0 +1,134 @@
+package tarsplitutils
+
+import (
+	"io"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/vbatts/tar-split/tar/storage"
+)
+
+func min(x, y int) int {
+	if x < y {
+		return x
+	}
+	return y
+}
+
+func max(x, y int) int {
+	if x < y {
+		return y
+	}
+	return x
+}
+
+type randomAccessTarStream struct {
+	entries storage.Entries
+	fg storage.FileGetter
+}
+
+func (self randomAccessTarStream) ReadAt(p []byte, off int64) (n int, err error) {
+	// Find the first entry that we're interested in
+	firstEntry := 0
+
+	cur_off := int64(0)
+	for i, entry := range self.entries {
+		var size int64
+
+		switch entry.Type {
+		case storage.SegmentType:
+			size = int64(len(entry.Payload))
+		case storage.FileType:
+			size = entry.Size
+		default:
+			return 0, fmt.Errorf("Unknown tar-split entry type: %v", entry.Type)
+		}
+
+		if cur_off <= off && off < cur_off + size {
+			firstEntry = i
+			break
+		}
+
+		cur_off += size
+	}
+
+	// The cursor will most likely be negative the first time. This signifies
+	// that we need to read some data first before starting to fill the buffer
+	n = int(cur_off - off)
```

Comment:
Here we calculate the difference of the current offset and offset and THEN casting so the two would have to be apart for more than 2GB. But that doesn't seem that improbable. I'll switch this declaration to int64 and then cast to `int` later when I know the difference is within the order of magnitude of `len(p)`